tlair/src/main/scala/at/forsyte/apalache/tla/lir/transformations/standard/ReTLALanguagePred.scala

```diff
@@ -0,0 +1,179 @@
+        PredResultOk()
+
+      case NameEx(_) =>
+        PredResultOk()
+
+      case OperEx(oper, arg) if ReTLALanguagePred.unaryOps.contains(oper) =>
+        isOkInContext(letDefs, arg)
+
+      case OperEx(oper, lhs, rhs) if ReTLALanguagePred.binaryOps.contains(oper) =>
+        isOkInContext(letDefs, lhs)
+          .and(isOkInContext(letDefs, rhs))
+
+      case OperEx(oper, args @ _*) if ReTLALanguagePred.naryOps.contains(oper) =>
+        allArgsOk(letDefs, args)
+
+      case OperEx(oper, NameEx(_), set, pred) if ReTLALanguagePred.bindingOps.contains(oper) =>
+        isOkInContext(letDefs, set).and(isOkInContext(letDefs, pred))
+
+      case OperEx(TlaControlOper.ifThenElse, pred, thenEx, elseEx) =>
+        isOkInContext(letDefs, pred)
+          .and(isOkInContext(letDefs, thenEx))
+          .and(isOkInContext(letDefs, elseEx))
+
+      case OperEx(TlaFunOper.funDef, args @ _*) =>
+        allArgsOk(letDefs, args)
+
+      // Function application and except are the only place we allow tuples, because that's how multivariable functions
+      // get expanded
+      case OperEx(TlaFunOper.app, fn, arg) =>
+        isOkInContext(letDefs, fn).and(
+            arg match {
+              case OperEx(TlaFunOper.tuple, args @ _*) => allArgsOk(letDefs, args)
+              case _                                   => isOkInContext(letDefs, arg)
+            },
+        )
+
+      // SANY quirk #19123:
+      // [f EXCEPT ![1] = ...] expands to
+      // OperEx( except, << 1 >>, ... ) and
+      // [f EXCEPT ![1,2] = ...] expands to
+      // OperEx(except, << << 1, 2 >> >> , ...)
+      // reTLA supports ONLY one-at-a-time, single-level EXCEPT, i.e.
+      // [f EXCEPT ![a,b,c,...] = r] (not [f EXCEPT ![a][b] = r] or [f EXCEPT ![a] = r, ![b] = s] )
+      case OperEx(TlaFunOper.except, fn, key, value) =>
+        isOkInContext(letDefs, fn)
+          .and(
+              isOkInContext(letDefs, value),
+          )
+          .and(
+              key match {
+                // ![a,b,...] case
+                case OperEx(TlaFunOper.tuple, OperEx(TlaFunOper.tuple, args @ _*)) =>
+                  allArgsOk(letDefs, args)
+                // ![a] case
+                case OperEx(TlaFunOper.tuple, args @ _*) =>
+                  allArgsOk(letDefs, args)
+                // Impossible, but we need case-completeness
+                case _ => PredResultFail(List())
+              },
+          )
+
+      case LetInEx(body, defs @ _*) =>
+        // go inside the let definitions (similar to FlatLanguagePred)
+        def eachDefRec(ctx: Set[String], ds: List[TlaOperDecl]): PredResult = {
+          ds match {
+            case Nil =>
+              PredResultOk()
+
+            case head :: tail =>
+              isOkInContext(ctx, head.body) // check the first operator definition
+                .and(eachDefRec(ctx + head.name, tail)) // check the other operator definitions
+          }
```

Comment:
Since even these comments are copied are from `KeraLanguagePred` (or one of it's ancestors or descendants), it does seem like some code sharing would be in order. If you don't like the `Foldable` suggestion, you could perhaps subclass these from a shared parent class?

Reply:
Yes, I suppose the helper method sharing makes sense.

Reply:
The helper sharing does indeed help a bit. Still more duplicated code here than I'd prefer, but don't want to block on that.